spec/mem/tree.md

```diff
@@ -0,0 +1,564 @@
+                Permission::Frozen => Permission::Disabled,
+                Permission::Disabled => Permission::Disabled,
+            }
+        )
+    }
+
+    fn permission_transition(curr: Permission, access_pattern: AccessKind, node_relation: NodeRelation) -> Result<Permission> {
+        match (access_pattern, node_relation) {
+            (AccessKind::Read, NodeRelation::Child) => Self::child_read(curr),
+            (AccessKind::Write, NodeRelation::Child) => Self::child_write(curr),
+            (AccessKind::Read, NodeRelation::Foreign) => Self::foreign_read(curr),
+            (AccessKind::Write, NodeRelation::Foreign) => Self::foreign_write(curr),
+        }
+    }
+}
+```
+
+We define the tree structure to track reborrows in each allocation. The tree consists of several nodes, each corresponding to a borrow tag (i.e., a reborrow). Each node has a parent and a list of children. Additionally, we track each tag's permissions on memory locations (bytes) within the allocation.
+
+```rust
+pub struct Tree {
+    root_tag: BorTag, 
+    /// Map a borrow tag to a node
+    nodes: Map<BorTag, Node>,
+}
+
+pub struct Node {
+    /// Borrow tag of the parent node
+    parent: Option<BorTag>, 
+    /// Borrow tags of the children node
+    children: List<BorTag>,
+    /// Permission for each location
+    permissions: Map<Address, Permission>,
+}
+
+pub enum NodeRelation {
+    Child, 
+    Foreign,
+}
+
+impl Tree {
+    /// Insert a node into the tree
+    fn insert_node(&mut self, tag: BorTag, node: Node) {
+        self.nodes.insert(tag, node);
+    }
+}
+```
+
+To update the tree during each memory access, each node in the tree can be divided into two disjoint sets: Child and Foreign. The Child set includes the node itself and all its descendants, while the Foreign set contains all other nodes. When a node is accessed, we must update and check permissions for each node in the tree based on its relative position to the accessed node.
+
+```rust
+impl TreeBorrowsAllocation {
+    /// Recusively do permission transition on the `curr` node and all its descendants.
+    /// `base` means the node that is actually memory accessed.
+    /// This method will throw UBs, representing the undefined behavior captured by Tree Borrows.
+    fn node_access(
+        &mut self, 
+        curr: BorTag,
+        access_pattern: AccessKind,
+        ptr: Pointer<TreeBorrowsProvenance>,
+        size: Size,
+        base: BorTag,
+    ) -> Result<bool> {
+        let Some(mut node) = self.tree.nodes.get(curr) else { throw_ub!("Tree Borrows: node not existed"); };
+
+        // Flag to indicate whether `base`is a child of `curr`
+        let mut is_child = curr == base;
+
+        for child_tag in node.children {
+            is_child |= self.node_access(child_tag, access_pattern, ptr, size, base)?;
+        }
+
+        let node_relation = if is_child { NodeRelation::Child } else { NodeRelation::Foreign };
```

Comment:
A comment should explain the direction of this relation -- i.e., who's the "self" here. It seems to be "curr"? I.e., "child" means that `base` is a child of `curr`.

Reply:
Yes, please document this everywhere, and have one central place where you put lots of examples. In our Coq development, this was not documented, and we did it the wrong way around an _embarrasing_ number of times.